babel/src/test/java/org/apache/calcite/test/BabelParserTest.java

```diff
@@ -2021,4 +2021,36 @@ private void checkParseInfixCast(String sqlType) {
     final String expected = "(?s).*Numeric literal.*out of range.*";
     sql(sql).fails(expected);
   }
+
+  @Test void testHexCharLiteralCharSetNotSpecifiedDefaultFormat() {
+    final String sql = "'c1a'XC";
+    final String expected = "'C1A' XC";
+    expr(sql).ok(expected);
+  }
+
+  @Test void testHexCharLiteralCharSetSpecifiedXCFormat() {
+    final String sql = "_KANJISJIS 'ABC'XC";
+    final String expected = "_KANJISJIS 'ABC' XC";
+    expr(sql).ok(expected);
+  }
+
+  @Test void testHexCharLiteralCharSetSpecifiedXCVFormat() {
+    final String sql = "_LATIN'c1a'XCV";
+    final String expected = "_LATIN 'C1A' XCV";
+    expr(sql).ok(expected);
+  }
+
+  @Test void testHexCharLiteralCharSetSpecifiedXCFFormat() {
+    final String sql = "_unicode'c1a'XCF";
+    final String expected = "_UNICODE 'C1A' XCF";
+    expr(sql).ok(expected);
+  }
+
+  @Test void testHexCharLiteralOutsideRangeFails() {
+    // 'g' contains char outside hex range, it would be incorrectly parsed
+    // falling into the <PREFIXED_STRING_LITERAL> which leads to errors
+    final String sql = "^_unicode'cg'^XCF";
+    final String expected = "Unknown character set 'unicode'";
+    expr(sql).fails(expected);
+  }
```

Comment:
As mentioned in design doc, I think it would also be good to test Hex Literals inside a query (ex. SELECT).

Reply:
Good point, will add a test case inside a query.